framework/include/samplers/Sampler.h

```diff
@@ -16,172 +16,204 @@
+   *
+   * Use these with caution as they can result in a large amount of memory use, the prefered
+   * method for accessing Stampler data is the getNextLocalRow() method.
    */
-  virtual void sampleSetUp(){};
-  virtual void sampleTearDown(){};
+  DenseMatrix<Real> getSamples();
+  DenseMatrix<Real> getLocalSamples();
   ///@}
 
   /**
-   * Return the sampled distribution data.
+   * Return the "next" local row. This is designed to be used within a loop using the
+   * getLocalRowBegin/End methods as such:
+   *
+   * for (dof_id_type i = getLocalRowBegin(); i < getLocalRowEnd(); ++i)
+   *     std::vector<Real> row = getNextLocalRow();
+   *
+   * This is the prefered method for accessing Sampler data.
+   *
+   * Calls to getNextLocalRow() will continue to return the next row of data until the last local
+   * row has been reached, it will then start again at the beginning if called again. Also, calls
+   * to getNextLocalRow() can be partial, followed by call(s) to getSamples or getLocalSamples.
+   * Continued calls to getNextLocalRow() will still continue to give the next row as if the
+   * other get calls were not made. However, when this occurs calls to restore and advance the
+   * generators are made after each call to getSamples or getLocalSamples, so this generally
+   * should be avoided.
    */
-  std::vector<DenseMatrix<Real>> getSamples();
+  std::vector<Real> getNextLocalRow();
 
   /**
-   * Return the sample names, by default 'sample_0, sample_1, etc.' is used.
-   * @return The names assigned to the DenseMatrix items returned by getSamples().
-   *
-   * If custom names are required then utilize the setSampleNames method.
+   * Return the number of samples.
+   * @return The total number of rows that exist in all DenseMatrix values from the
+   * getSamples/getLocalSamples methods.
    */
-  const std::vector<std::string> & getSampleNames() const { return _sample_names; }
+  dof_id_type getNumberOfRows() const;
+  dof_id_type getNumberOfCols() const;
+  dof_id_type getNumberOfLocalRows() const;
 
+  ///@{
   /**
-   * Set the sample names.
-   * @param names List of names to assign to the DenseMatrix entries returned by
-   *              getSamples() method.
+   * Return the beginning/end local row index for this processor
    */
-  void setSampleNames(const std::vector<std::string> & names);
+  dof_id_type getLocalRowBegin() const;
+  dof_id_type getLocalRowEnd() const;
+  ///@}
 
+protected:
+  // The following methods are the basic methods that should be utilized my most application
+  // developers that are creatign a custom Sampler.
+
+  ///@{
   /**
-   * Store the state of the MooseRandom generator so that new calls to getSamples will create
-   * new numbers.
+   * These methods must be called within the constructor of child classes to define the size of
+   * the matrix to be created.
    */
-  void execute();
+  void setNumberOfRows(dof_id_type n_rows);
```

Comment:
@aeslaughter In my case, I depend on a UserObject to compute the number of samples, and I can't do the `getUserObject()` call in the sampler's constructor. If I had an `initialSetup()` method I could call `setNumberOfRows()` there.

Reply:
Could `Sampler` derive from `SetupInterface`?

Reply:
As discussed, I will try to move init() to the first Sampler::execute() call.

Reply:
@bwspenc That doesn't work. It needs to happen in the constructor to allow for the correct number of MultiApp objects to be initialized. We are going to need to think about this more.